tiledb/common/evaluator/evaluator.h

```diff
@@ -0,0 +1,249 @@
+    typename List::iterator lru_it;
+    std::shared_ptr<Value> value;
+    Cached_Entry() = default;
+  };
+
+  /**
+   * Enforces the cache policy by evicting entries if necessary.
+   * This operation assumes the bookkeeping mutex is already locked.
+   */
+  virtual std::shared_ptr<Value> enforce_policy(size_t mem_usage) = 0;
+
+  /**
+   * Double linked list where the head is the least
+   * recently used element from the cache.
+   */
+  List lru_;
+
+  /**
+   * Each entry stores a promised value which is in the process of
+   * being fetched. Readers will wait on it to become available.
+   */
+  std::unordered_map<Key, OndemandValue> inprogress_;
+
+  /**
+   * Map of cached entries by key. Each entry stores and iterator
+   * within the LRU list and a shared_ptr to the value.
+   * If an entry is in this map, it is guaranteed that the value is
+   * already available to serve readers.
+   */
+  std::unordered_map<Key, Cached_Entry> entries_;
+
+  /** Bookkeeping lock */
+  std::mutex mutex_;
+};
+
+template <class Key, class Value>
+class NoCachePolicy {
+ public:
+  using key_type = Key;
+  using value_type = Value;
+
+  NoCachePolicy() = default;
+
+  template <class F>
+  std::shared_ptr<Value> operator()(F f, Key key) {
+    return std::get<1>(f(key));
+  }
+};
+
+template <class Key, class Value, size_t N = 0>
+class MaxEntriesCache : public CachePolicyBase<Key, Value> {
+ public:
+  MaxEntriesCache()
+      : num_entries_(0) {
+  }
+
+  template <class F>
+  std::shared_ptr<Value> operator()(F f, Key key) {
+    return std::get<1>(f(key));
+  }
+
+  virtual std::shared_ptr<Value> enforce_policy(
+      [[maybe_unused]] size_t mem_usage) override {
+    if (num_entries_ == max_entries_) {
+      auto evicted_value = this->evict();
+      --num_entries_;
+      return evicted_value;
+    }
+    ++num_entries_;
+    return nullptr;
+  }
+
+ private:
+  size_t num_entries_;
+  static const size_t max_entries_ = N;
+};
+
+template <class Key, class Value, size_t N = 1024>
+class MemoryBudgetedCache : public CachePolicyBase<Key, Value> {
+ public:
+  MemoryBudgetedCache()
+      : memory_consumed_(0) {
+  }
+
+  template <class F>
+  std::shared_ptr<Value> operator()(F f, Key key) {
+    return std::get<1>(f(key));
+  }
+
+  virtual std::shared_ptr<Value> enforce_policy(size_t mem_usage) override {
+    if (memory_consumed_ + mem_usage > memory_budget_) {
+      auto evicted_value = this->evict();
+      memory_consumed_ -= mem_usage;
+      return evicted_value;
+    }
+
+    memory_consumed_ += mem_usage;
+    return nullptr;
+  }
+
+ private:
+  /** The maximum amount of bytes managed by this cache */
+  static const size_t memory_budget_ = N;
+
+  /** The maximum amount of bytes currently consumed by this cache */
+  size_t memory_consumed_;
+};
+
+template <class Policy, class F>
+class Evaluator {
+ public:
+  using Key = typename Policy::key_type;
+  using Value = typename Policy::value_type;
+
+  Evaluator()
```

Comment:
It's unlikely that all the policies will only ever need zero arguments. Setting up argument forwarding now would be prudent.

As an example, memory budgets would be best passed by value here so that calls to `Config::get` are outside the policy definition.